### PR TITLE
rados_utils_5x.py: Adding common rados functionality

### PR DESCRIPTION
**Pull request by pdhiran**
Moving all the commonly used rados functionality in one place

Signed-off-by: Pawan Dhiran <pdhiran@redhat.com>

### CODE REVIEW

**Comment by pdhiran** [comment]
http://magna002.ceph.redhat.com/ceph-qe-logs/pawan/cephci-run-1627549031139/

File: ceph/rados/rados_utils_5x.py [hunk 1]
```diff
@@ -0,0 +1,791 @@
...
+        cmd = f"sudo rados --no-log-to-stderr -b {byte_size} -p {pool_name} bench {duration} write --no-cleanup"
+        try:
+            self.node.shell([cmd])
+            return True
+        except Exception as err:
+            log.error(f"Error running rados bench write on pool : {pool_name}")
+            log.error(err)
+            return False
+
+    def run_rados_bench_read(self, pool_name: str, **kwargs) -> bool:
```
Comment (psathyan) [inline ceph/rados/rados_utils_5x.py:432]: Likewise... we could have the method name as `bench_read`

**Review by psathyan — COMMENTED** [review]
I strongly believe we are heading in the right direction. The workflows defined here looks re-usable by upper layer. Let's improve the documentation

File: ceph/rados/rados_utils_5x.py [hunk 2]
```diff
@@ -0,0 +1,791 @@
...
+import logging
+import re
+import time
+
+from ceph.ceph_admin import CephAdmin
+
+log = logging.getLogger(__name__)
+
+
+class RadosTools:
```
[the remark this answers is not in the document]
Comment (pdhiran) [reply ceph/rados/rados_utils_5x.py]: Changed the name that describes the use of object.

File: ceph/rados/rados_utils_5x.py [hunk 3]
```diff
@@ -0,0 +1,791 @@
+"""
```
[the remark this answers is not in the document]
Comment (pdhiran) [reply ceph/rados/rados_utils_5x.py]: changed the file to `core_workflows.py`

File: ceph/rados/rados_utils_5x.py [hunk 4]
```diff
@@ -0,0 +1,791 @@
...
+                command = f"{cfg_map[cmd]} {config.get(cmd, 8)}"
+            else:
+                command = cfg_map[cmd]
+            self.node.shell([command])
+
+    def reweight_crush_items(self, **kwargs) -> bool:
+        """
+        Performs Re-weight of various CRUSH items, based on key-value pairs sent
+        Args:
+            **kwargs: Arguments for the commands
```
[the remark this answers is not in the document]
Comment (pdhiran) [reply ceph/rados/rados_utils_5x.py]: Added supported KW pairs for all the methods which required them.

File: ceph/rados/rados_utils_5x.py [hunk 5]
```diff
@@ -0,0 +1,791 @@
...
+                )
+                # Sleeping for 5 seconds after command execution for process to start
+                time.sleep(5)
+            else:
+                log.info(
+                    "No re-weights based on utilization were triggered. PG distribution is optimal"
+                )
+                return True
+
+        if kwargs.get("verify_reweight"):
```
[the remark this answers is not in the document]
Comment (pdhiran) [reply ceph/rados/rados_utils_5x.py]: Added new method to verify the re-weight of OSD's

File: ceph/rados/rados_utils_5x.py [hunk 6]
```diff
@@ -0,0 +1,791 @@
...
+            return True
+        log.error(f"Pool:{pool} could not be deleted on cluster")
+        return False
+
+    def configure_pg_autoscaler(self, **kwargs) -> bool:
+        """
+        Configures pg_Autoscaler as a global global parameter and on pools
+        Args:
+            node: Cephadm node where the commands need to be executed
+            **kwargs: Any other param that needs to be set
```
[the remark this answers is not in the document]
Comment (pdhiran) [reply ceph/rados/rados_utils_5x.py]: Added the supported KW pairs

File: ceph/rados/rados_utils_5x.py [hunk 7]
```diff
@@ -0,0 +1,791 @@
...
+
+        """
+        cmd = "ceph config set mgr mgr/cephadm/config_checks_enabled true"
+        self.node.shell([cmd])
+
+        # Checking if the checks are enabled on cluster
+        cmd = "ceph cephadm config-check status"
+        out, err = self.node.shell([cmd])
+        if not re.search("Enabled", out):
+            log.error("Cluster config checks no t enabled")
```
[the remark this answers is not in the document]
Comment (pdhiran) [reply ceph/rados/rados_utils_5x.py]: Corrected the typo.

File: ceph/rados/rados_utils_5x.py [hunk 8]
```diff
@@ -0,0 +1,791 @@
...
+        if kwargs.get("default_mode"):
+            default_mode = kwargs.get("default_mode")
+            cmd = f"ceph config set global osd_pool_default_pg_autoscale_mode {default_mode}"
+            self.node.shell([cmd])
+
+        cmd = "ceph osd pool autoscale-status -f json"
+        log.info(self.node.shell([cmd]))
+        return True
+
+    def set_cluster_configuration_checks(self, **kwargs) -> bool:
```
[the remark this answers is not in the document]
Comment (pdhiran) [reply ceph/rados/rados_utils_5x.py]: added methods to do the same.

File: ceph/rados/rados_utils_5x.py [hunk 9]
```diff
@@ -0,0 +1,791 @@
...
+            self.node.shell([cmd])
+        except Exception as err:
+            log.error("Error while Sending email alerts")
+            log.error(err)
+            return False
+
+        log.info("Email alerts configured on the cluster")
+        return True
+
+    def run_rados_bench_write(self, pool_name: str, **kwargs) -> bool:
```
[the remark this answers is not in the document]
Comment (pdhiran) [reply ceph/rados/rados_utils_5x.py]: changed

File: ceph/rados/rados_utils_5x.py (hunk 1 above)
Comment (pdhiran) [reply ceph/rados/rados_utils_5x.py]: changed.

File: ceph/rados/rados_utils_5x.py [hunk 10]
```diff
@@ -0,0 +1,791 @@
...
+        if kwargs.get("enable_check_list"):
+            for check in kwargs.get("enable_check_list"):
+                cmd = f"ceph cephadm config-check enable {check}"
+                self.node.shell([cmd])
+
+        cmd = "ceph cephadm config-check ls"
+        log.info(self.node.shell([cmd]))
+        return True
+
+    def set_logging_to_file(self) -> bool:
```
[the remark this answers is not in the document]
Comment (pdhiran) [reply ceph/rados/rados_utils_5x.py]: changed.

**Review by psathyan — APPROVED** [review]
Looks good to me

**Pull request merged**